lineapy/plugins/pipeline_writers.py

```diff
@@ -0,0 +1,206 @@
+
+        logger.info("Generated Docker file")
+
+    def write_pipeline_files(self) -> None:
+        self._write_dag()
+        self._write_docker()
+
+
+class AirflowPipelineWriter(BasePipelineWriter):
+    """
+    Class for pipeline file writer. Corresponds to "AIRFLOW" framework.
+    """
+
+    def __init__(
+        self,
+        session_artifacts_sorted: List[SessionArtifacts],
+        keep_lineapy_save: bool,
+        pipeline_name: str,
+        output_dir: str,
+    ) -> None:
+        self.session_artifacts_sorted = session_artifacts_sorted
+        self.keep_lineapy_save = keep_lineapy_save
+        self.pipeline_name = pipeline_name
+        self.output_dir = Path(output_dir, pipeline_name)
+
+        # Create output directory folder(s) if nonexistent
+        self.output_dir.mkdir(exist_ok=True, parents=True)
+
+        # We assume there is at least one SessionArtifacts object
+        self.db = self.session_artifacts_sorted[0].db
+
+    def _write_dag(
+        self,
+        airflow_dag_config: Optional[AirflowDagConfig] = {},
+        airflow_dag_flavor: str = "PythonOperatorPerSession",
+    ) -> None:
+        airflow_dag_config = airflow_dag_config or {}
+        if (
+            AirflowDagFlavor[airflow_dag_flavor]
+            == AirflowDagFlavor.PythonOperatorPerSession
+        ):
+            AIRFLOW_DAG_TEMPLATE = load_plugin_template(
+                "airflow_dag_PythonOperatorPerSession.jinja"
+            )
+            session_functions = [
+                f"run_session_including_{session_artifacts._get_first_artifact_name()}"
+                for session_artifacts in self.session_artifacts_sorted
+            ]
+            dependencies = {
+                session_functions[i + 1]: {session_functions[i]}
+                for i in range(len(session_functions) - 1)
+            }
+            task_graph = TaskGraph(
+                nodes=session_functions,
+                mapping={f: f for f in session_functions},
+                edges=dependencies,
+            )
+            full_code = AIRFLOW_DAG_TEMPLATE.render(
+                DAG_NAME=self.pipeline_name,
+                MODULE_NAME=self.pipeline_name + "_module",
+                OWNER=airflow_dag_config.get("owner", "airflow"),
+                RETRIES=airflow_dag_config.get("retries", 2),
+                START_DATE=airflow_dag_config.get("start_date", "days_ago(1)"),
+                SCHEDULE_INTERVAL=airflow_dag_config.get(
+                    "schedule_interval", "*/15 * * * *"
+                ),
+                MAX_ACTIVE_RUNS=airflow_dag_config.get("max_active_runs", 1),
+                CATCHUP=airflow_dag_config.get("catchup", "False"),
+                tasks=session_functions,
+                task_dependencies=task_graph.get_airflow_dependency(),
+            )
+            full_code = prettify(full_code)
+
+            # Write out file
+            file = self.output_dir / f"{self.pipeline_name}_dag.py"
```

Comment:
Can we pass output_dir and pipeline_name in as a parameter to the function? It will make it more testable in the future to have all function dependencies passed in.

Reply:
i prefer the existing flow here. these writers are used internally and the output dirs / names etc are defined outside the writer classes. Having to pass them again during write will mean the code using these writers need to remember when to pass what params which i dont like. as for testing ( these are class methods ), its easy to init a class instance with custom paths/names so i'm indifferent there.